### ISSUE TITLE
import numpy fails with pypy3 and numpy 1.15 or 1.16: undefined symbol: PyGILState_Check

### ISSUE THREAD

**Issue by seanlaguna** [numpy/numpy#13152]
I installed pypy3 from the Gentoo Project:Prefix (https://wiki.gentoo.org/wiki/Project:Prefix) repositories with `emerge pypy3`. It is up-to-date with the latest pypy3 (7.1.0-alpha0 is the version given). I'm on an older Red Hat machine so this is why I use Project:Prefix. I don't think this is relevant to the problem but who knows. I installed numpy inside a virtualenv using pip. I do have other versions of python installed but inside the virutalenv it shouldn't see those. I did not build numpy from source, unless using pip does that automatically somehow.

### Reproducing code example:

```
virtualenv -p ~/gentoo/usr/bin/pypy3 pypy3test
source pypy3test/bin/activate
pip install numpy
python
>>>> import numpy
```

### Error message:

```
[slaguna] ~/venv
$ python
Python 3.5.3 (1f86f25937b6ae6c8b25236c35228fac587678bf, Mar 16 2019, 17:19:57)
[PyPy 7.1.0-alpha0 with GCC 8.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
And now for something completely different: ``:-) or rather ":-/"
off-by-one-key error''
>>>> import numpy
Traceback (most recent call last):
  File "/data/slaguna/venv/pypy3test/site-packages/numpy/core/__init__.py", line 40, in <module>
    from . import multiarray
  File "/data/slaguna/venv/pypy3test/site-packages/numpy/core/multiarray.py", line 12, in <module>
    from . import overrides
  File "/data/slaguna/venv/pypy3test/site-packages/numpy/core/overrides.py", line 6, in <module>
    from numpy.core._multiarray_umath import (
ImportError: /data/slaguna/venv/pypy3test/site-packages/numpy/core/_multiarray_umath.pypy3-71-x86_64-linux-gnu.so: undefined symbol: PyGILState_Check

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/data/slaguna/venv/pypy3test/site-packages/numpy/__init__.py", line 142, in <module>
    from . import core
  File "/data/slaguna/venv/pypy3test/site-packages/numpy/core/__init__.py", line 71, in <module>
    raise ImportError(msg)
ImportError: 

IMPORTANT: PLEASE READ THIS FOR ADVICE ON HOW TO SOLVE THIS ISSUE!

Importing the multiarray numpy extension module failed.  Most
likely you are trying to import a failed build of numpy.
Here is how to proceed:
- If you're working with a numpy git repository, try `git clean -xdf`
  (removes all files not under version control) and rebuild numpy.
- If you are simply trying to use the numpy version that you have installed:
  your installation is broken - please reinstall numpy.
- If you have already reinstalled and that did not fix the problem, then:
  1. Check that you are using the Python you expect (you're using /data/slaguna/venv/pypy3test/bin/python),
     and that you have no directories in your PATH or PYTHONPATH that can
     interfere with the Python and numpy versions you're trying to use.
  2. If (1) looks fine, you can open a new issue at
     https://github.com/numpy/numpy/issues.  Please include details on:
     - how you installed Python
     - how you installed numpy
     - your operating system
     - whether or not you have multiple versions of Python installed
     - if you built from source, your compiler versions and ideally a build log

     Note: this error has many possible causes, so please don't comment on
     an existing issue about this - open a new one instead.

Original error was: /data/slaguna/venv/pypy3test/site-packages/numpy/core/_multiarray_umath.pypy3-71-x86_64-linux-gnu.so: undefined symbol: PyGILState_Check
```

### Numpy/Python version information:

```
[slaguna] ~/venv
$ python
Python 3.5.3 (1f86f25937b6ae6c8b25236c35228fac587678bf, Mar 16 2019, 17:19:57)
[PyPy 7.1.0-alpha0 with GCC 8.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
And now for something completely different: ``happy PyPy's year 2010!''
>>>> import sys; print(sys.version)
3.5.3 (1f86f25937b6ae6c8b25236c35228fac587678bf, Mar 16 2019, 17:19:57)
[PyPy 7.1.0-alpha0 with GCC 8.2.0]
```

**Comment by mattip** [comment]
Issues with support for PyPy should be raised on their [issue tracker](https://bitbucket.org/pypy/pypy/issues) first. It seems PyPy does not implement `PyGILState_Check`, which they should do.

**Comment by seanlaguna** [comment]
No problem, I'll report to their issue tracker. I was just following the directions in the error message by reporting here. Can you let me know if there's a way to compile numpy without the need for this function? It looks like it only occurs in one place inside an `#ifdef` block, and I know other people have successfully run numpy with pypy3.

**Comment by mattip** [comment]
I am not sure. On ubuntu 18.04 I can download the [latest nightly python3.6 build](http://buildbot.pypy.org/nightly/py3.6) and run 
```
pypy3 -mensurepip
pypy3 -mpip install pytest cython
pypy3 runtests.py
```
which runs the complete numpy test suite. There are failures, but not the one you report.

**Comment by mattip** [comment]
perhaps because the calls to `assert(NPY_CHECK_GIL_HELD())` where `NPY_CHECK_GIL_HELD` is defined as `PyGILState_Check` are inside asserts, and the normal compilation of NumPy eliminates the asserts by defining `NDEBUG`

**Comment by rth** [comment]
For future reference the upstream issue is [pypy#3041](https://bitbucket.org/pypy/pypy/issues/3041/pypy3-import-numpy-fails-unknown-symbol) and it looks like it was fixed in https://bitbucket.org/pypy/pypy/commits/0fcdbc3de300 on the py3.6 branch. I'm not sure if this implies if it's fixed in pypy 7.x under Python 3.6. Will try it out.

**Comment by mattip** [comment]
This will be fixed in the next release of PyPy. Since it is an API change, the version will be 8.x

**Comment by rth** [comment]
Thanks for the clarification!

> While NumPy has a CI run with latest PyPy3, it does not run in debug mode so this line is not hit. You can rerun the build with `CFLAGS=-DNDEBUG`.

Incidentally I also run into this issue on Gentoo Linux, maybe it's something about that distribution that doesn't set `-DNDEBUG`. I also had to manually set `-O2` to avoid https://github.com/numpy/numpy/issues/14137 with v1.17.0. Or maybe I changed the default CFLAGS in some config files and don't remember anymore...
